### DISCUSSION TITLE
Non Boolean Interpolation Results?

### DISCUSSION THREAD

**Discussion by baierd** [cvc5/cvc5#8996] [Q&A] [answered]
Hello CVC5 Team,

i am currently trying to incorporate CVC5 into [JavaSMT](https://github.com/sosy-lab/java-smt).
Since JavaSMT is very interested in Craig-Interpolation we were especially interested in CVC5, since, according to your recent paper, CVC5 should be capable of producing Craig-Interpolants.
We are using the Java API. And i can provide examples if need be.

I've now tried using the CVC5 interpolation API, but could only produce Boolean results.
I tried getting the next interpolant also, but either recieve a Boolean result again, or the null Term.
Is there a option i am missing? (Besides the option to enable interpolation: produce-interpolants true)
Or are Boolean results simply the norm for CVC5?

Thank you for your help!
Best,
Daniel

**Comment by yoni206** [comment]
Hi Daniel, 

Thank you for opening this discussion.
Would it be possible for you to provide an example of the issue?
Thanks!

**Comment by baierd** [comment]
Sure. Written in Java, one of the examples we use with all interpolating solvers:

```
    solver = new Solver();
    // Set the logic
    solver.setLogic("ALL");

    // options we use
    solver.setOption("incremental", "true");
    solver.setOption("produce-models", "true");
    solver.setOption("finite-model-find", "true");
    solver.setOption("sets-ext", "true");
    solver.setOption("output-language", "smtlib2");
    solver.setOption("strings-exp", "true");
    solver.setOption("produce-interpolants", "true");

    Term zero = solver.mkInteger(0);
    Term one = solver.mkInteger(1);

    Term a = solver.mkConst(solver.getIntegerSort(), "a");
    Term b = solver.mkConst(solver.getIntegerSort(), "b");
    Term c = solver.mkConst(solver.getIntegerSort(), "c");

    // build formula:  1 = A = B = C = 0
    Term aEq1 = solver.mkTerm(Kind.EQUAL, one, a);
    Term aEqb = solver.mkTerm(Kind.EQUAL, a, b);
    Term bEqc = solver.mkTerm(Kind.EQUAL, b, c);
    Term cEq0 = solver.mkTerm(Kind.EQUAL, c, zero);

    solver.assertFormula(aEq1);
    solver.assertFormula(aEqb);
    solver.assertFormula(bEqc);
    solver.assertFormula(cEq0);

    assertThat(solver.checkSat().isUnsat()).isTrue();

    Term itp = solver.getInterpolant(solver.mkBoolean(true));
    Term itpA = solver.getInterpolant(aEq1);
    Term itpAB = solver.getInterpolant(aEq1.andTerm(aEqb));
    Term itpABC = solver.getInterpolant(aEq1.andTerm(aEqb).andTerm(bEqc));
    Term itpD = solver.getInterpolant(cEq0);
    Term itpDC = solver.getInterpolant(cEq0.andTerm(bEqc));
    Term itpDCB = solver.getInterpolant(cEq0.andTerm(bEqc).andTerm(aEqb));
    Term itpABCD = solver.getInterpolant(aEq1.andTerm(aEqb).andTerm(bEqc).andTerm(cEq0));

    System.out.println(itp);
    System.out.println(itpA);
    System.out.println(itpAB);
    System.out.println(itpABC);
    System.out.println(itpD);
    System.out.println(itpDC);
    System.out.println(itpDCB);
    System.out.println(itpABCD);
```

Prints:
```
true
false
false
false
false
false
false
false
```

Expected printout would be something like this (same example, just SMTInterpol als a solver):
```
true
(<= 0 (+ a (- 1)))
(<= 0 (+ b (- 1)))
(<= 0 (+ c (- 1)))
(<= c 0)
(<= b 0)
(<= a 0)
false
```

**Comment by yoni206** [comment]
Thanks for all these details!

I might be missing something here, but it seems like the second call to `getInterpolant` searches for a formula that is implied by the conjunction  `aEq1 /\ aEqb /\ bEqc /\cEq0`, and implies `aEq1 `. 

`false` is indeed such a formula: It is implied by the conjunction (this conjunction is by itself inconsistent, as it implies `1=0`), and it implies every formula, in particular `aEq1`.

There might be other such formulas, but `cvc5` searches for the smallest one. If the conjunction (set of formulas that were asserted to `cvc5` before calling `getInterpoland`) is consistent, perhaps more "interesting" interpolants can be obtained.

I hope that helps, and will be happy to assist and continue the discussion.

**Comment by ajreynol** [comment]
You can try the smt2 command `get-interpolant-next` / API method `Solver::getInterpolantNext` if you want cvc5 to return another candidate interpolant.  This should produce an interpolant that is different than the current one.

**Comment by baierd** [comment, marked as answer]
Thank you for your replies!
Is the kind of interpolation used by CVC5 [Craig-Interpolation](https://doi.org/10.2307/2963593) according to the following definition?

For a pair of formulas φ− and φ+ such that φ− ∧ φ+ is
unsatisfiable, a Craig interpolant ψ is a formula that fulfills
the following requirements:
1) the implication φ− ⇒ ψ holds,
2) the conjunction ψ ∧ φ+ is unsatisfiable, and
3) ψ only contains symbols that occur in both φ− and φ+.

If yes, the API documentation of the Solver::getInterpolant() states that the assertion stack is A (φ− ) while the input into getInterpolant() is B (φ+). However some tests by my side have shown that that does not seem to be the case.
If we take the example from above and modify it like so:

```
    solver = new Solver();
    // Set the logic
    solver.setLogic("ALL");

    // options we use
    solver.setOption("incremental", "true");
    solver.setOption("produce-models", "true");
    solver.setOption("finite-model-find", "true");
    solver.setOption("sets-ext", "true");
    solver.setOption("output-language", "smtlib2");
    solver.setOption("strings-exp", "true");
    solver.setOption("produce-interpolants", "true");

    Term zero = solver.mkInteger(0);
    Term one = solver.mkInteger(1);

    Term a = solver.mkConst(solver.getIntegerSort(), "a");
    Term b = solver.mkConst(solver.getIntegerSort(), "b");
    Term c = solver.mkConst(solver.getIntegerSort(), "c");

    // build formula:  1 = A = B = C = 0
    Term aEq1 = solver.mkTerm(Kind.EQUAL, one, a);
    Term aEqb = solver.mkTerm(Kind.EQUAL, a, b);
    Term bEqc = solver.mkTerm(Kind.EQUAL, b, c);
    Term cEq0 = solver.mkTerm(Kind.EQUAL, c, zero);

    solver.assertFormula(aEq1);
    Term itpA = solver.getInterpolant(aEqb.andTerm(bEqc).andTerm(cEq0));
```

then itpA should produce a valid interpolant here. However only null is produced.
(Hence why i interpreted that the assertion stack is really A ∧ B (φ− ∧ φ+ from the definition))

**Reply by yoni206** [reply]
Thanks for this example and question.

```
1. the implication φ− ⇒ ψ holds,
2. the conjunction ψ ∧ φ+ is unsatisfiable, and
```

Not this variant, but another, according to which $\phi^{-}\rightarrow \psi$ and $\psi\rightarrow\phi^{+}$ are both valid. 

I hope the semantics of `getInterpolant` is clearer. This is also described in `getInterpolant` description here:
https://cvc5.github.io/docs/cvc5-1.0.1/api/java/io/github/cvc5/Solver.html#getInterpolant(io.github.cvc5.Term)

Please let us know if you have any more questions / issues.

**Reply by baierd** [reply]
Yes this clears up things! Thank you very much for your help!

**Reply by baierd** [reply]
@yoni206 i have a followup question regarding your answer. Could you specify what you mean by `valid`?
We had a discussion about this interpolation procedure and iff ψ→ϕ+ holds for all possible assignments, your procedure and our definition are actually interchangeable. Did you mean to say with `valid` that the implications have to hold for all assignments?

**Reply by yoni206** [reply]
When the logic is arithmetic (like in the example above) -- yes.
For other logics, it is not only for all assignments, but also for all underlying models (in arithmetic there is only one underlying model).
